Ensure -k is set to delegated hosts without a pass

jborean93 opened the pull request:
##### SUMMARY
The change in https://github.com/ansible/ansible/pull/69244 stopped combining the original vars with the delegated host vars. This meant certain variables set by the play context on the host was not transferred to the delegated host, like `ansible_password`. This change reimplements the combining of connection vars to replicate the original behaviour.

Fixes https://github.com/ansible/ansible/issues/71103

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
task_executor

jborean93 committed:
Ensure -k is set to delegated hosts without a pass

jborean93 commented:
@bcoca it would be great if you could have a look over this change. https://github.com/ansible/ansible/pull/69244 swapped around the original logic to fix another bug and it seems like cvars is used exclusively for `_set_connection_options()` which did `combine_vars` before.

jborean93 committed:
Fix up some broken tests

bcoca commented on lib/ansible/executor/task_executor.py:
```diff
@@ -507,7 +507,8 @@ def _execute(self, variables=None):
             # use vars from delegated host (which already include task vars) instead of original host
-            cvars = variables.get('ansible_delegated_vars', {}).get(self._task.delegate_to, {})
+            delegate_vars = variables.get('ansible_delegated_vars', {}).get(self._task.delegate_to, {})
+            cvars = combine_vars(variables, delegate_vars)
```
no, combine vars with inventory_hostname variables is specifically what we want to avoid, this does not only bring the `-k` defined password but any that where set in inventory for the inventory_hostname, then applying possibly wrong password (among other settings) to the delegated_to host.  What we do need to do is pull the play_context passwords  (in the end they should really be applied to the task, but that is future fix) and let delegated_to host override if set.

bcoca replied:
need to x2 check, but `  self._play_context.update_vars(cvars)` shoudl do the trick here

bcoca replied:
currently other plugins use play_context directly as a fallback ` conn_password = self.get_option('password') or self._play_context.password`

bcoca replied:
after thinking for a bit, bets place should be to add passwords to the `set_options` calls in  task_executor, both for connection and become by merging into 'task_keys' before passing it in (at lowest priority) which will allow settings/vars to keep the right precedence and insure its always present

bcoca replied:
https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py#L903

bcoca replied:
actually .. it seems we do attempt to do that .. could it be we have a bug in that code?

bcoca replied:
ah ... winrm uses `remote_password` not `password` .. why it is not getting picked up.

bcoca replied:
not sure how we want to fix it, add 'remote_password' as another task field or update the winrm plugin?

samalv replied:
may I ask please why WinRM uses remote_password and not password ?

jborean93 replied:
@samalv they are just keyword names for the config options which they never really were used for anything special before. I mostly followed the same standard for `remote_user` and `remote_addr` but after this change it does have a meaning. I've added the alias for `password` and fixed the code to actually use the aliases when setting the options for keywords.

samalv replied:
@jborean93 thanks for explanation, should we expect this fix to be pushed in next Ansible release 2.9.12 ? or shall I do the same fix manually on my installed version 2.9.11 ?

jborean93 replied:
I'll be creating a backport for this so the next 2.9 release will include it. Unfortunately it is too late for 2.10.0 so it will be in the 2.10.1 release later on.

samalv replied:
yes I saw them, what about 2.11 ?

bcoca committed:
Update task_executor.py

one possible fix, the other is updating winrm to normalize on 'password' like the other connection plugins

jborean93 committed:
Add alias for winrm and fix incorrect assumption

jborean93 committed:
Make sure aliases are used for keyword options

jborean93 committed:
Conditionally run test if sshpass is present, fix sanity

jborean93 commented:
Backport PRs

stable-2.9 - https://github.com/ansible/ansible/pull/71167
stable-2.10 - https://github.com/ansible/ansible/pull/71168
